File: ui/src/api/endpoints/material.ts
/**
 * Material Management API Endpoints
 */
import apiClient from '../client';

// Types
export interface FeederSetup {
  setup_id: string;
  line_code: string;
  feeder_slot: string;
  feeder_type?: string;
  material_code?: string;
  material_name?: string;
  reel_id?: string;
  lot_no?: string;
  initial_qty: number;
  remaining_qty: number;
  status: 'active' | 'empty' | 'error' | 'standby';
  setup_time?: string;
  setup_by?: string;
  created_at: string;
}

export interface MaterialConsumption {
  consumption_id: string;
  order_id?: string;
  line_code: string;
  equipment_code?: string;
  material_code: string;
  material_name?: string;
  lot_no?: string;
  reel_id?: string;
  consumption_qty: number;
  unit: string;
  consumption_type: 'normal' | 'loss' | 'scrap' | 'rework';
  consumption_time: string;
  product_code?: string;
  result_lot_no?: string;
  created_at: string;
}

export interface MaterialRequest {
  request_id: string;
  request_no?: string;
  request_type: 'replenish' | 'return' | 'transfer';
  line_code: string;
  equipment_code?: string;
  feeder_slot?: string;
  material_code: string;
  material_name?: string;
  requested_qty: number;
  unit: string;
  urgency: 'normal' | 'urgent' | 'critical';
  status: 'requested' | 'approved' | 'in_transit' | 'delivered' | 'cancelled';
  requested_by?: string;
  requested_at: string;
  approved_by?: string;
  approved_at?: string;
  delivered_by?: string;
  delivered_at?: string;
  delivered_qty?: number;
  remarks?: string;
  created_at: string;
}

export interface MaterialInventory {
  inventory_id: string;
  location_type: 'line' | 'buffer' | 'supermarket' | 'wip';
  location_code: string;
  material_code: string;
  material_name?: string;
  lot_no?: string;
  qty_on_hand: number;
  qty_allocated: number;
  qty_available: number;
  unit: string;
  min_qty: number;
  max_qty: number;
  last_count_date?: string;
  last_count_qty?: number;
  created_at: string;
}

export interface LineMaterialSummary {
  line_code: string;
  line_name?: string;
  total_feeders: number;
  active_feeders: number;
  empty_feeders: number;
  error_feeders: number;
  pending_requests: number;
  urgent_requests: number;
}

export interface MaterialShortageAlert {
  line_code: string;
  feeder_slot: string;
  material_code: string;
  material_name?: string;
  remaining_qty: number;
  estimated_empty_time?: string;
  urgency: string;
}

// Feeder Setup APIs
export const getFeederSetups = async (params?: {
  line_code?: string;
  status?: string;
  material_code?: string;
  page?: number;
  page_size?: number;
}) => {
  const { data } = await apiClient.get('/api/v1/mes/material/feeders', { params });
  return data;
};

export const getLineFeeders = async (lineCode: string) => {
  const { data } = await apiClient.get<FeederSetup[]>(`/api/v1/mes/material/feeders/${lineCode}`);
  return data;
};

export const createFeederSetup = async (feederData: Partial<FeederSetup>) => {
  const { data } = await apiClient.post<FeederSetup>('/api/v1/mes/material/feeders', feederData);
  return data;
};

export const updateFeederSetup = async (setupId: string, feederData: Partial<FeederSetup>) => {
  const { data } = await apiClient.patch<FeederSetup>(`/api/v1/mes/material/feeders/${setupId}`, feederData);
  return data;
};

// Material Consumption APIs
export const getMaterialConsumption = async (params?: {
  line_code?: string;
  material_code?: string;
  start_date?: string;
  end_date?: string;
  page?: number;
  page_size?: number;
}) => {
  const { data } = await apiClient.get('/api/v1/mes/material/consumption', { params });
  return data;
};

export const createMaterialConsumption = async (consumptionData: Partial<MaterialConsumption>) => {
  const { data } = await apiClient.post<MaterialConsumption>('/api/v1/mes/material/consumption', consumptionData);
  return data;
};

// Material Request APIs
export const getMaterialRequests = async (params?: {
  line_code?: string;
  status?: string;
  urgency?: string;
  page?: number;
  page_size?: number;
}) => {
  const { data } = await apiClient.get('/api/v1/mes/material/requests', { params });
  return data;
};

export const createMaterialRequest = async (requestData: Partial<MaterialRequest>) => {
  const { data } = await apiClient.post<MaterialRequest>('/api/v1/mes/material/requests', requestData);
  return data;
};

export const updateMaterialRequest = async (requestId: string, requestData: Partial<MaterialRequest>) => {
  const { data } = await apiClient.patch<MaterialRequest>(`/api/v1/mes/material/requests/${requestId}`, requestData);
  return data;
};

// Material Inventory APIs
export const getMaterialInventory = async (params?: {
  location_code?: string;
  material_code?: string;
  low_stock?: boolean;
  page?: number;
  page_size?: number;
}) => {
  const { data } = await apiClient.get('/api/v1/mes/material/inventory', { params });
  return data;
};

export const createMaterialInventory = async (inventoryData: Partial<MaterialInventory>) => {
  const { data } = await apiClient.post<MaterialInventory>('/api/v1/mes/material/inventory', inventoryData);
  return data;
};

export const updateMaterialInventory = async (inventoryId: string, inventoryData: Partial<MaterialInventory>) => {
  const { data } = await apiClient.patch<MaterialInventory>(`/api/v1/mes/material/inventory/${inventoryId}`, inventoryData);
  return data;
};

// Summary & Alerts APIs
export const getLineMaterialSummary = async () => {
  const { data } = await apiClient.get<LineMaterialSummary[]>('/api/v1/mes/material/summary/lines');
  return data;
};

export const getMaterialShortageAlerts = async () => {
  const { data } = await apiClient.get<MaterialShortageAlert[]>('/api/v1/mes/material/alerts/shortage');
  return data;
};

File: api/schemas/mes/material.py
"""
MES Material Management Schemas
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from enum import Enum
from pydantic import BaseModel, Field
from uuid import UUID


# Enums
class FeederType(str, Enum):
    TAPE_8MM = "tape_8mm"
    TAPE_12MM = "tape_12mm"
    TAPE_16MM = "tape_16mm"
    TAPE_24MM = "tape_24mm"
    TRAY = "tray"
    STICK = "stick"


class FeederStatus(str, Enum):
    ACTIVE = "active"
    EMPTY = "empty"
    ERROR = "error"
    STANDBY = "standby"


class ConsumptionType(str, Enum):
    NORMAL = "normal"
    LOSS = "loss"
    SCRAP = "scrap"
    REWORK = "rework"


class RequestType(str, Enum):
    REPLENISH = "replenish"
    RETURN = "return"
    TRANSFER = "transfer"


class RequestUrgency(str, Enum):
    NORMAL = "normal"
    URGENT = "urgent"
    CRITICAL = "critical"


class RequestStatus(str, Enum):
    REQUESTED = "requested"
    APPROVED = "approved"
    IN_TRANSIT = "in_transit"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class LocationType(str, Enum):
    LINE = "line"
    BUFFER = "buffer"
    SUPERMARKET = "supermarket"
    WIP = "wip"


# Feeder Setup Schemas
class FeederSetupBase(BaseModel):
    line_code: str
    feeder_slot: str
    feeder_type: Optional[FeederType] = None
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    reel_id: Optional[str] = None
    lot_no: Optional[str] = None
    initial_qty: int = 0
    remaining_qty: int = 0


class FeederSetupCreate(FeederSetupBase):
    equipment_id: Optional[UUID] = None
    setup_by: Optional[str] = None


class FeederSetupUpdate(BaseModel):
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    reel_id: Optional[str] = None
    lot_no: Optional[str] = None
    remaining_qty: Optional[int] = None
    status: Optional[FeederStatus] = None


class FeederSetupResponse(FeederSetupBase):
    setup_id: UUID
    equipment_id: Optional[UUID] = None
    setup_time: Optional[datetime] = None
    setup_by: Optional[str] = None
    status: FeederStatus = FeederStatus.ACTIVE
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Material Consumption Schemas
class MaterialConsumptionBase(BaseModel):
    line_code: str
    material_code: str
    material_name: Optional[str] = None
    lot_no: Optional[str] = None
    consumption_qty: Decimal
    unit: str = "EA"
    consumption_type: ConsumptionType = ConsumptionType.NORMAL


class MaterialConsumptionCreate(MaterialConsumptionBase):
    order_id: Optional[UUID] = None
    equipment_code: Optional[str] = None
    reel_id: Optional[str] = None
    product_code: Optional[str] = None
    result_lot_no: Optional[str] = None


class MaterialConsumptionResponse(MaterialConsumptionBase):
    consumption_id: UUID
    order_id: Optional[UUID] = None
    equipment_code: Optional[str] = None
    reel_id: Optional[str] = None
    product_code: Optional[str] = None
    result_lot_no: Optional[str] = None
    consumption_time: datetime
    created_at: datetime

    class Config:
        from_attributes = True


# Material Request Schemas
class MaterialRequestBase(BaseModel):
    request_type: RequestType = RequestType.REPLENISH
    line_code: str
    material_code: str
    material_name: Optional[str] = None
    requested_qty: Decimal
    unit: str = "EA"
    urgency: RequestUrgency = RequestUrgency.NORMAL


class MaterialRequestCreate(MaterialRequestBase):
    equipment_code: Optional[str] = None
    feeder_slot: Optional[str] = None
    requested_by: Optional[str] = None
    remarks: Optional[str] = None


class MaterialRequestUpdate(BaseModel):
    status: Optional[RequestStatus] = None
    approved_by: Optional[str] = None
    delivered_by: Optional[str] = None
    delivered_qty: Optional[Decimal] = None
    remarks: Optional[str] = None


class MaterialRequestResponse(MaterialRequestBase):
    request_id: UUID
    request_no: Optional[str] = None
    equipment_code: Optional[str] = None
    feeder_slot: Optional[str] = None
    status: RequestStatus = RequestStatus.REQUESTED
    requested_by: Optional[str] = None
    requested_at: datetime
    approved_by: Optional[str] = None
    approved_at: Optional[datetime] = None
    delivered_by: Optional[str] = None
    delivered_at: Optional[datetime] = None
    delivered_qty: Optional[Decimal] = None
    remarks: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# Material Inventory Schemas
class MaterialInventoryBase(BaseModel):
    location_type: LocationType = LocationType.LINE
    location_code: str
    material_code: str
    material_name: Optional[str] = None
    lot_no: Optional[str] = None
    qty_on_hand: Decimal = Decimal("0")
    qty_allocated: Decimal = Decimal("0")
    unit: str = "EA"


class MaterialInventoryCreate(MaterialInventoryBase):
    min_qty: Decimal = Decimal("0")
    max_qty: Decimal = Decimal("0")


class MaterialInventoryUpdate(BaseModel):
    qty_on_hand: Optional[Decimal] = None
    qty_allocated: Optional[Decimal] = None
    min_qty: Optional[Decimal] = None
    max_qty: Optional[Decimal] = None


class MaterialInventoryResponse(MaterialInventoryBase):
    inventory_id: UUID
    qty_available: Decimal = Decimal("0")
    min_qty: Decimal = Decimal("0")
    max_qty: Decimal = Decimal("0")
    last_count_date: Optional[datetime] = None
    last_count_qty: Optional[Decimal] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Summary Schemas
class LineMaterialSummary(BaseModel):
    line_code: str
    line_name: Optional[str] = None
    total_feeders: int = 0
    active_feeders: int = 0
    empty_feeders: int = 0
    error_feeders: int = 0
    pending_requests: int = 0
    urgent_requests: int = 0


class MaterialShortageAlert(BaseModel):
    line_code: str
    feeder_slot: str
    material_code: str
    material_name: Optional[str] = None
    remaining_qty: int
    estimated_empty_time: Optional[datetime] = None
    urgency: str
